### ISSUE TITLE
What culture does the built-in SQL function UPPER and LOWER use?

### ISSUE THREAD

**Issue by Liversage** [Azure/azure-documentdb-dotnet#409]
To perform case insensitive queries in Cosmos DB I can use the SQL function `UPPER` to convert a string to upper case. Unfortunately, the documentation does not specify how this conversion is done. Different cultures does it in different ways.

A good test case for case conversion is the lower case dotless I `"ı"` and the capital dotted I `"İ"`. These letters are used in the Turkish language so if you have Turkish names in your database you can run into problems if the conversion to upper case is done differently in different parts of your query.

Assume that I have a document with a `name` property and I want to do a case insensitive search for users by name. In this example I'm looking for persons that has the dotless I in their name so I enter this letter in the front-end.

It is possible to craft the query so `UPPER` is used on the JSON property and the value to compare with:

    SELECT * FROM c WHERE CONTAINS(UPPER(c.name), UPPER("ı"))

This search will work reliably even though it may give more results than the user expected because `UPPER` will convert both the dotless `"ı"` and the dotted `"i"` to `"I"` so the query will also find people with `"i"` in their name. That is OK.

However, if you are using .NET to interface with Cosmos DB using LINQ you can run into problems if you are not careful. The query above can be written using LINQ:

    client
        .CreateDocumentQuery<User>(...)
        .Where(user => user.Name.ToUpper().Contains(searchTerm.ToUpper());

Let's assume that `searchTerm` is a lower case dotless I as above. This will be converted into SQL with the following structure:

    SELECT * FROM c WHERE CONTAINS(UPPER(c.name), "ı"))

Notice that the LINQ provider converts `searchTerm` to upper case before inserting the value into the generated SQL. This is quite sensible. However, depending on the culture of the C# process the conversion to upper case will vary. If the culture is `CultureInfo.InvariantCulture` then the lower case dotless I will actually not be converted (as in the example above). For most other cultures including `en-US` it will be converted to a capital (dotless) I (at least that is my experience).

Because you cannot reliably predict how `ToUpper` converts a letter (it depends on the environment) it is best practice to always specify a `CultureInfo` or use `ToUpperInvariant` (which is equivalent to specifying `CultureInfo.InvariantCulture`). So in my actual code where I encountered this issue I used `searchTerm.ToUpperInvariant()` which resulted in the lower case dotless I being inserted into the SQL as shown above.

An because `UPPER("ı")` results in `"I"` while `"ı".ToUpperInvariant()` results in `"ı"` the query does not give me the results I want (people with `ı` in the name).

Summing up...

The issue is that when using .NET and LINQ to query Cosmos DB with the use of `ToUpper` you need to know the exact details of how `UPPER` works in Cosmos DB SQL to be able to perform the same conversion to upper case in the .NET code because the LINQ provider will execute as much code as it can on the client side resulting in upper case conversion being done in two slightly different ways.

Can you provide these case conversion details in the documentation so even Turkish names can be found when doing case insensitive searches?

**Comment by Liversage** [comment]
Perhaps my issue is written in a very complicated way. I would like to rephrase it into this hopefully simpler question that should have a clear answer:

I have a user collection with documents that have a `name` property. I want to perform queries for parts of the name in a case insensitive manner. I do this by converting the name to upper case. I'm using C# and LINQ:

    var queryable = client
        .CreateDocumentQuery<User>(...)
        .Where(user => user.Name.ToUpper().Contains(searchTerm.ToUpper(cultureInfo));

The `user.Name.ToUpper()` is translated to the Cosmos DB SQL function `UPPER` while `searchTerm.ToUpper(cultureInfo)` is evaluated in .NET and inserted as a literal string in the SQL query.

I need to know which `cultureInfo` to supply to `ToUpper` in C# to ensure that the conversion done in .NET is the same as the conversion done in Cosmos DB by `UPPER`. I have discovered that `CultureInfo.InvariantCulture` is not the right culture to use because it does not convert a lower case dotless `"ı"` in the same way as Cosmos DB does.

**Comment by bchong95** [comment]
Thank you for the detailed description. 

This is a very fair ask and we have an internal work item for this. I will update the thread if the work item has updates.

**Comment by DegJ** [comment]
I have just run into an issue for just this.
Example:
Data in Cosmos:
`{
  "City": "Örebro"
}`

If I now use the .net client to do a search with lowercasing (using a searchquery "ö" in his case (already in lower, no conversion needed)):
`.Where(x=> x.City.ToLower().Contains("ö".ToLower())`
It will not match any document (same with trying with uppercase "Ö" in searchquery.
It seems to not be working when using non english characters like "ÅÄÖ". But using a query with a english character like "B"
`.Where(x=> x.City.ToLower().Contains("B".ToLower())`
Will match the above document

**Comment by Liversage** [comment]
@bchong95 I just noticed your edit:

> The short answer is that the BUILT-INs do not use a cultureInfo, so it is whatever the defaults are.

My question is basically what is the default in Cosmos DB? I assume that Cosmos DB is not built using .NET so "what `CultureInfo` Cosmos DB uses" is not a good question. However, it must still use some specific rules for `UPPER` and `LOWER` and I'm asking about the details of these rules so I can use the same rules in .NET code.

I guess a solution to my specific LINQ problem would be to ensure that the Cosmos DB LINQ provider translates both calls to `ToUpper()` into `UPPER` in Cosmos DB SQL instead of only one being translated while the other executes in my .NET process leading to this problem with some non-ASCII characters.
